enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,964 @@
+
+```bash 
+$ opm index inspect --index=docker.io/my-namespace/community-operators --output=json
+$ cd community-operators
+$ tree
+.
+├── etcd.json
+├── amqstreams.json
+├── image_sha256.json
+```
+#### Editing a package inside an index
+
+A new sub-command `update` will be introduced under `opm index`, which will take a folder containing json representations of packages as it's input along with an existing index image tag, and will push any edit made to the config files up to the index container image. 
+
+```bash
+$ opm index inspect --index=quay.io/some-namespace/my-index 
+$ // edit a my-index/package-config.json 
+$ opm index update my-index --tag=docker.io/some-namespace/my-index:latest
+```
+
+When an index is "inspected" for it's config files, a `image_sha256.json` will also be created that will store the sha of the index image being inspected. 
+```bash 
+$ cat community-operators/image_sha256.json
+{
+    "sha256":"9e0c9c2c8e973c7e17439509d1dc1512afa3a8ed7c0b169016299a357bbc5b10"
+}
+```
+When an update is attempted on an index image, the `update` command will verify that the sha information locally matches with the sha for the remote index image before overriding the content of the container image with the latest changes. This will prevent overriding of existing information that may not have been pulled locally due to asynchronous update of the config files. 
+
+> Note: For a seamless UX, the following things should be consider during implementation when the `opm index update` command is summoned: 
+> 1. Validate the files inside the folder given as input to the command to ensure they are valid json representation of packages (and not random files with random content, which could be potential security risks)
+> 2. If the files are valid files, verify that the sha information stored locally matches the sha of the container image from the registry it's hosted on. If they do not match, throw an error. 
+> 3. If the files are valid, and the sha is valid, take a git diff between the present content of the index (i.e the existing json representations) and the content of the folder given as input. Display the diff as an output of the command with the following message 
+> "The following changes will be updated in the index:"
+> If there is no diff between the content in the container image and the content of the local folder, throw an error with the message "No diff found between index and local content to update."    
+
+
+#### Creating new indexes using the config files from an existing index
+
+A new sub-command `create` will be introduced under `opm index`, that will take a folder containing json/yaml representation of packages, and a container image tag as input, and will create a new index with the package representations fed to it. 
+
+```bash
+$ tree community-operators 
+community-operators
+├── amqstreams.json
+└── etcd.json
+$ opm index create --from=community-operators --tag=docker.io/some-namespace/community-operators
```

Comment:
It's not explicitly stated, but I assume that this will walk a tree of files, find all json files, and slurp them up?

Reply:
Made it more explicit in the first sentence of this section.

Reply:
I think it's also important to also have a way to continue building the Index Image scratch.  The `opm registry create` is no longer useful here, but something akin to `opm index create ... --generate`.

With the intermix of upstream/downstream `opm`, choosing the correct version of the `opm` builder image as well as multi-architecture images has been problematic.

Reply:
Agreed. We need to have an easy way to determine which `opm` version should be used for a catalog of a certain schema version.